### PR TITLE
XYChart: Add data filter to manual mode

### LINKED ISSUE #80651: XYChart: Updates to manual mode
1. When data source has multiple frames and a field name is common across them, currently displays them all on a single series - prevent this.
2. Add data frame selection per series to make this mode fully manual
3. Point color calcs are per series, add an option to make color calcs global across all series

### PR DESCRIPTION
**Pull request by drew08t**
Updates manual mode to include a data filter. When no data filter is selected, all fields are available to pick.

- [x] Add data filter
- [x] Store data filter selection per series
- [x] Handle clearing data filter
- [x] Trigger update to UI upon data filter change
- [x] Apply data filter to FieldNamePickers
- [x] Ensure uniqueness for frame filter (fields with same display name across frames)
- [x] Force data filter to be chosen per series - can default to first frame to make selection easier

Closes #80651

### CODE REVIEW

**Comment by drew08t** [comment]
There is a known limitation for manual mode currently, which I'm considering following up on or trying to address here:

If you have two queries with fields sharing the same names, two series, one with a data filter and one with no filter (both using the same fields for x and y), series generation incorrectly creates 3 series (both rendering and in the legend).
Jan-26-2024 08-52-45

**Comment by drew08t** [comment]
> There is a known limitation for manual mode currently, which I'm considering following up on or trying to address here:
> 
> If you have two queries with fields sharing the same names, two series, one with a data filter and one with no filter (both using the same fields for x and y), series generation incorrectly creates 3 series (both rendering and in the legend). Jan-26-2024 08-52-45 [ Jan-26-2024 08-52-45 ](https://private-user-images.githubusercontent.com/60050885/300050988-ed801dea-97a0-4903-b06c-2d2188d011ea.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY1NDQ0MDksIm5iZiI6MTcwNjU0NDEwOSwicGF0aCI6Ii82MDA1MDg4NS8zMDAwNTA5ODgtZWQ4MDFkZWEtOTdhMC00OTAzLWIwNmMtMmQyMTg4ZDAxMWVhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI5VDE2MDE0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1YzkyNTlhOGY0NjI2ZjFjMTRhOWNiMGM1NTViOGIxZTIzYmI1ZTIwNDIwNjZlNGNiYmE3YTI1ZTZiZWM1ODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0MNOimrD5s-s0HV6RN5rMdTNZBfPqY2G3aFCwoqZ8o0) [ ](https://private-user-images.githubusercontent.com/60050885/300050988-ed801dea-97a0-4903-b06c-2d2188d011ea.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY1NDQ0MDksIm5iZiI6MTcwNjU0NDEwOSwicGF0aCI6Ii82MDA1MDg4NS8zMDAwNTA5ODgtZWQ4MDFkZWEtOTdhMC00OTAzLWIwNmMtMmQyMTg4ZDAxMWVhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI5VDE2MDE0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1YzkyNTlhOGY0NjI2ZjFjMTRhOWNiMGM1NTViOGIxZTIzYmI1ZTIwNDIwNjZlNGNiYmE3YTI1ZTZiZWM1ODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0MNOimrD5s-s0HV6RN5rMdTNZBfPqY2G3aFCwoqZ8o0)

The decision has been made to force the user to use data filter per series for manual mode, making it fully explicit as one would expect.

File: public/app/plugins/panel/xychart/ManualEditor.tsx [hunk 1]
```diff
@@ -101,23 +112,53 @@ export const ManualEditor = ({
...
+                  frameNames.find((v) => {
+                    return v.value === value[selected].frame;
+                  }) ?? 0
                 }
-                return obj;
-              })
-            );
-          }}
-        />
+                onChange={(v) => {
```
Comment (baldm0mma) [inline public/app/plugins/panel/xychart/ManualEditor.tsx:127]: Nit: I love when callback function args are typed, just for quick/glancing readability/context.

File: public/app/plugins/panel/xychart/ManualEditor.tsx [hunk 2]
```diff
@@ -101,23 +112,53 @@ export const ManualEditor = ({
...
+                    })
+                  );
+                }}
+              />
+            </Field>
+          )}
+          <ScatterSeriesEditor
+            key={`series/${selected}`}
+            baseNameMode={FieldNamePickerBaseNameMode.ExcludeBaseNames}
+            item={{} as StandardEditorsRegistryItem}
```
Comment (baldm0mma) [inline public/app/plugins/panel/xychart/ManualEditor.tsx:146]: I thought type assertions were naughtynaughty? 🤔

File: public/app/plugins/panel/xychart/ManualEditor.tsx [hunk 3]
```diff
@@ -101,23 +112,53 @@ export const ManualEditor = ({
...
+            context={context}
+            value={value[selected]}
+            onChange={(v) => {
```
Comment (baldm0mma) [inline public/app/plugins/panel/xychart/ManualEditor.tsx:149]: same nit here. :)

**Comment by baldm0mma** [comment]
Is it worth adding any tests?

**Review by baldm0mma — APPROVED** [review]
Big fan of your work. 😃

File: public/app/plugins/panel/xychart/ManualEditor.tsx (hunk 2 above)
Comment (nmarrs) [reply public/app/plugins/panel/xychart/ManualEditor.tsx:146]: typically this object is used to provide any editor specific settings as well as the `id` which is used to persist the information in configs and `name` which can be changed and doesn't impact configs

in this case we don't have any specific settings for this editor and we embedding a custom editor within another custom editor (vs using builder api) so I think this is acceptable 🤷🏻‍♂️

the alternative would be something like `item={{ id: `series/${selected}`, name: `series/${selected}` }}` but I don't feel strongly either way on this

**Review by nmarrs — APPROVED** [review]
overall this LGTM and makes sense - there is some strange UX / the x / y selection breaks when disabling / removing queries (not found - this is expected though so 🤷) and the filter only updates on data refresh but pure functionality wise this works

[test xy chart-1706654226053.json](https://github.com/grafana/grafana/files/14105122/test.xy.chart-1706654226053.json)


https://github.com/grafana/grafana/assets/22381771/f5fba97e-09fb-4b08-adee-8a990a341f2e



Screenshot 2024-01-30 at 2 34 05 PM

File: public/app/plugins/panel/xychart/ManualEditor.tsx [hunk 4]
```diff
@@ -18,6 +19,16 @@ export const ManualEditor = ({
   onChange,
   context,
 }: StandardEditorProps<ScatterSeriesConfig[], unknown, Options>) => {
+  const frameNames = useMemo(() => {
+    if (context?.data?.length) {
```
[the remark this answers is not in the document]
Comment (drew08t) [reply public/app/plugins/panel/xychart/ManualEditor.tsx:23]: length of 0 should return false, no?

[the remark this answers is not in the document]
Comment (nmarrs) [reply public/app/plugins/panel/xychart/ManualEditor.tsx:23]: yeah that's true - we can leave as is 👍

[the remark this answers is not in the document]
Comment (baldm0mma) [reply public/app/plugins/panel/xychart/ManualEditor.tsx:23]:
```
const noLengthString = ""
console.log(noLengthString.length)
// 0
```

[the remark this answers is not in the document]
Comment (nmarrs) [reply public/app/plugins/panel/xychart/ManualEditor.tsx:23]: object will either be `undefined` or a `number`, if `undefined` will be considered `false` and if `0` will be considered false

**Comment by drew08t** [comment]
> Is it worth adding any tests?

Yes! Let's do it! https://github.com/grafana/grafana/issues/81608

**Pull request merged**